First try

DenisAlimov opened the pull request.

DenisAlimov committed:
Second commit

DenisAlimov committed:
Добавляется сущность-связь

DenisAlimov committed:
Некоторые правки существующего фунционала

DenisAlimov committed:
Добавил обработку некоторых исключений

DenisAlimov committed:
Исправил выброс ошибки по добавлению книг. Добавил некоторые аннотации для Swagger

DenisAlimov committed:
Починил зависимости для корректной работы Swagger. Убрал лишние неиспользуемые зависимости. Добавил настройки для docker/docker-compose и сохранения jar в корне проекта.

EvdokimovaIrina commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,19 @@
+spring.sql.init.data-locations=classpath:init.sql
+spring.sql.init.mode=always
+
+spring.jpa.hibernate.ddl-auto=create-drop
```
Не очень хорошая практика разрешать хиберу самому создавать схему тут лучше указать только валидировать (spring.jpa.hibernate.ddl-aut=validate), ну и не хватает скриптов по созданию таблиц. По заданию "К проекту необходимо предоставить скрипт для создания таблиц и наполнения тестовыми данными в виде файла init.sql"

EvdokimovaIrina commented on src/main/java/com/example/Library/controllers/BookController.java:
```diff
@@ -0,0 +1,56 @@
+
+    @ApiOperation("method to get list of authors")
+    @GetMapping("/authors")
+    public List<AuthorResponse> getAllAuthors() {
```
Этот endpoint лучше было вынести в отдельный контроллер, переделывать не надо, просто как комментарий

EvdokimovaIrina replied:
не хватает входных параметров
GET /authors – получение списка авторов  
Данный endpoint должен возвращать список авторов и количество книг у каждого с использованием пагинации (сортировка пагинации по authorFullName по возрастанию)
Входные параметры:
int page - номер страницы
int size - количество элементов на странице

EvdokimovaIrina commented on src/main/java/com/example/Library/facade/BookDataFacade.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.stream.Collectors;
+
+@Component
+public class BookDataFacade {
```
Лишний компонент, все это надо делать в слое сервиса тут же можно только DTO  менять, но в данном случае это лишнее, а то получается что классы сервиса ни чего по сути не делают, а всю логику сбора данных делает этот класс

EvdokimovaIrina commented on src/main/java/com/example/Library/facade/BookDataFacade.java:
```diff
@@ -0,0 +1,96 @@
+                .map(author -> AuthorResponse.builder()
+                        .authorId(author.getId())
+                        .authorFullName(author.getAuthorFullName())
+                        .booksQty(bookAuthorService.countBookQuantity(author))
```
Так не подойдет, плохой подход ходить отдельно в БД для каждого автора, а если их будет 1 млн? Надо этот список получать одним запросом

EvdokimovaIrina commented on src/main/java/com/example/Library/facade/BookDataFacade.java:
```diff
@@ -0,0 +1,96 @@
+                .map(book -> BookResponse.builder()
+                        .bookId(book.getId())
+                        .bookName(book.getBookName())
+                        .authors(bookAuthorService.getAuthorsByBookId(book.getId()))
```
Аналогично получению авторов, надо получать одним запросом, тем более что в сущность Book ты добавил List<BookAuthor> bookAuthor.

EvdokimovaIrina commented on src/main/java/com/example/Library/facade/BookDataFacade.java:
```diff
@@ -0,0 +1,96 @@
+            throw new ValidationException("Book should have at least one author");
+        }
+
+        BookDto bookDto = new BookDto(bookAuthorRequest.getBookName());
```
А зачем тут из BookAuthorRequest в BookDto, а потом в Book? Ну и автора надо бы проверить, может уже есть такой, а если нет, то создавать и id автора не передается, а получается при добавлении, а потом уже добавляется в книгу и все коммититься. Не хватает @Transactional особенно при добавлении

EvdokimovaIrina replied:
Посмотрела в задании указано в запросе  "authorId", но он не нужен или если оставлять, то надо проверять, если в запросе передан authorId, то значит искать по нему автора ну и еще проверку ФИО делать, если нет, то добавлять нового и у него генерируется свой id

EvdokimovaIrina commented on src/main/java/com/example/Library/service/WikiServiceImpl.java:
```diff
@@ -0,0 +1,39 @@
+    public String getWiki(int id) throws JsonProcessingException {
+
+        String bookName = bookService.getBookById(id).getBookName();
+        String resourceUrl = "https://ru.wikipedia.org/w/api.php?action=query&list=search&srsearch=" + bookName + "&utf8=&format=json";
```
комментарий на будущее, желательно вынести url в property, ну и  проще использовать @FeignClient

EvdokimovaIrina commented on src/main/java/com/example/Library/controllers/BookController.java:
```diff
@@ -0,0 +1,56 @@
+        return bookDataFacade.createBookWithAuthors(bookAuthorRequest);
+    }
+
+    @GetMapping("/books")
```
Не хватает входных параметров. По заданию 
GET /books – получение списка книг 
Данный endpoint должен возвращать список книг и их авторов с использованием пагинации (сортировка пагинации по bookName по возрастанию)
Входные параметры:
int page - номер страницы
int size - количество элементов на странице

EvdokimovaIrina commented on src/main/java/com/example/Library/response/BookResponse.java:
```diff
@@ -0,0 +1,20 @@
+public class BookResponse implements Comparable<BookResponse> {
+    private int bookId;
+    private String bookName;
+    private List<Author> authors;
```
List<Author> тут по хорошему надо отдельную выходную DTO, как для BookResponse

EvdokimovaIrina commented on src/main/java/com/example/Library/response/BaseWebResponse.java:
```diff
@@ -0,0 +1,10 @@
+
+@Data
+@AllArgsConstructor
+public class BaseWebResponse {
```
Т.к. эта DTO относится к ошибкам, то в названии тоже лучше указать error, например BaseErrorResponse

DenisAlimov committed:
Добавил скрипты таблиц, заменил create-drop на validate.
Заменил RestTemplate на feign. Создал сущности под json.
Исправил название класса на BaseErrorResponse.

DenisAlimov committed:
Убрал неиспользуемые классы/пакеты/методы.
Исправил методы поиска авторов и книг, добавив пагинацию (сортировка+параметры, при ответе использовал запрос из БД с сортировкой по требуемым параметрам, воспользовался List<BookAuthor> вместо запросов из БД по каждому автору/книге.

DenisAlimov committed:
Убрал фасад, отрефакторил метод добавления книги с авторами, убрал лишние зависимости маппера, поправил ридми.

DenisAlimov committed:
Поправил аннотацию на @Transactional, добавил CascadeType в сущность Book, убрал неиспользуемый BookAuthorService. Добавление авторов сделал всем списком через saveAll.
Добавил несколько json для POST-запросов.

DenisAlimov committed:
Убрал лишнее каскадное добавление из сущности автора.

DenisAlimov committed:
Разкоментировал @Transactional!

DenisAlimov committed:
Убрал лишние импорты, комментарии, неиспользуемые методы, отформатировал код. Также добавил выброс ошибки, что книги с таким названием может не быть в википедии.

DenisAlimov committed:
Убрал неиспользуемые методы из BookAuthorRepository и добавил ссылку на Swagger в READMY.

DenisAlimov committed:
Вернул проверку на существующего автора, добавил тест createBook.

DenisAlimov committed:
Добавил проверку на лист автора, который передается. Осталось разобраться с книгой

DenisAlimov committed:
Сделал проверку на добавление книги!

DenisAlimov committed:
Нашел этот bookId!!!

DenisAlimov committed:
Нерабочий вариант с проверкой книги!

DenisAlimov committed:
Перенес создание книги в сервис, закомментировал неработающий тест, заменил его на any(Book.class).

DenisAlimov committed:
Добавил ленивую загрузку!
